src/auth.rs

```diff
@@ -138,12 +156,229 @@ where
     Ok(())
 }
 
+async fn authenticate_replit<SocketStream, WebSocketStream>(
+    stream: &mut SocketStream,
+    ws_stream: &mut WebSocketStream,
+    replid: &str,
+    pubkeys: &HashMap<String, Vec<u8>>,
+) -> Result<()>
+where
+    SocketStream: tokio::io::AsyncRead + tokio::io::AsyncWrite + Unpin + Send,
+    WebSocketStream: futures::Sink<WebSocketMessage, Error = tokio_tungstenite::tungstenite::error::Error>
+        + futures::Stream<
+            Item = std::result::Result<
+                WebSocketMessage,
+                tokio_tungstenite::tungstenite::error::Error,
+            >,
+        > + Unpin
+        + Send,
+{
+    client_handshake_step(ws_stream, b"RFB 003.008\n", b"RFB 003.008\n")
+        .await
+        .context("client ProtocolVersion handshake")?;
+
+    // Only support the VeNCrypt authentication.
+    client_handshake_step(ws_stream, b"\x01\x13", b"\x13")
+        .await
+        .context("client security type")?;
+
+    // Only support VeNCrypt 0.2.
+    client_handshake_step(ws_stream, b"\x00\x02", b"\x00\x02")
+        .await
+        .context("client VeNCrypt version")?;
+
+    // Only support Plain authentication.
+    ws_stream
+        .send(WebSocketMessage::Binary(b"\x00".to_vec()))
+        .await?;
+    client_handshake_step(ws_stream, b"\x01\x00\x00\x01\x00", b"\x00\x00\x01\x00")
+        .await
+        .context("client VeNCrypt subtype")?;
+
+    // All the preamble is done, now receive username+password.
+    let (username, password) = client_username_password(ws_stream).await?;
+    log::debug!("->: {:?} {:?}", &username, &password);
+
+    match validate_token(&username, &replid, pubkeys).context("token validation") {
+        Ok(()) => {}
+        Err(err) => {
+            // Let the client know that it messed up.
+            stream.write_all(b"\x00\x00\x00\x01").await?;
+            return Err(err);
+        }
+    };
+
+    // Now that the token itself was validated, we perform the handshake against the upstream RFB
+    // server.
+    server_handshake_step(stream, b"RFB 003.008\n", b"RFB 003.008\n")
+        .await
+        .context("server ProtocolVersion handshake")?;
+
+    let mut buf = [0u8; 1024];
+
+    // SecurityType handshake.
+    let mut n = stream.read(&mut buf[..]).await?;
+    if n <= 1 || buf[0] as usize + 1 != n {
+        bail!("invalid SecurityType payload {:?}", &buf[..n]);
+    }
+    log::debug!("<-: {:?}", &buf[..n]);
+
+    if buf[1..n].contains(&2) {
```

Comment:
Is this checking for a single `\x02` byte in a series of bytes like `\x00\x00\x02\x00`? Why can't this check for a specific byte?

Reply:
yup! each byte is the ID of a different securityType: https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#security, the protocol does not mandate that they are in any specific order, so we can't know a priori which byte to check.

let me add a comment about this.